sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,165 @@
+                return
+                    IsAllowedArrayType(type) ||
+                    IsAllowedListType(type) ||
+                    IsAllowedDictionaryType(type);
+            }
+
+            private static bool IsAllowedArrayType(Type type)
+            {
+                if (!type.IsArray)
+                {
+                    return false;
+                }
+
+                Type? elementType = type.GetElementType();
+                return elementType != null && IsAllowedType(elementType);
+            }
+
+            private static bool IsAllowedListType(Type type)
+            {
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (type.GetGenericTypeDefinition() != typeof(List<>))
+                {
+                    return false;
+                }
+
+                Type[] types = type.GetGenericArguments();
+                return IsAllowedType(types[0]);
+            }
+
+            private static bool IsAllowedDictionaryType(Type type)
+            {
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (type.GetGenericTypeDefinition() == typeof(Dictionary<,>))
+                {
+                    Type[] types = type.GetGenericArguments();
+                    if (types[0] == typeof(string) && IsAllowedType(types[1]))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private static bool IsAllowedAnonymousType(Type type, HashSet<Type> ancestorTypes)
+            {
+                if (!IsAnonymousType(type))
+                {
+                    return false;
+                }
+
+                foreach (PropertyInfo property in type.GetProperties())
+                {
+                    if (IsAllowedPrimitive(property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    if (IsAllowedCollectionType(property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    // Detect cycles: trust but verify
+                    if (ancestorTypes.Contains(property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    // Recurse
+                    ancestorTypes.Add(type);
+                    if (!IsAllowedAnonymousType(property.PropertyType, ancestorTypes))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private static bool IsAnonymousType(Type type)
+            {
+                return
+                    type.Namespace == null &&
+                    Attribute.IsDefined(type, typeof(CompilerGeneratedAttribute), false);
```

Comment:
How sure are we this check does not let some unintended types through?

Reply:
No - I will use Jared's suggested approach - I hadn't used it before because I was trying to avoid a string comparison.